src/hammer-tech/stackup.py

```diff
@@ -71,41 +73,44 @@ class WidthSpacingTuple(NamedTuple('WidthSpacingTuple', [
     __slots__ = ()
 
     @staticmethod
-    def from_setting(d: dict) -> "WidthSpacingTuple":
+    def from_setting(grid_unit: Decimal, d: dict) -> "WidthSpacingTuple":
         # pylint: disable=missing-docstring
-        width_at_least = float(d["width_at_least"])
-        min_spacing = float(d["min_spacing"])
-        assert width_at_least >= 0.0
-        assert min_spacing > 0.0
+        width_at_least = coerce_to_grid(d["width_at_least"], grid_unit)
+        min_spacing = coerce_to_grid(d["min_spacing"], grid_unit)
+        assert width_at_least >= 0
+        assert min_spacing > 0
         return WidthSpacingTuple(
             width_at_least=width_at_least,
             min_spacing=min_spacing
         )
 
+
     @staticmethod
-    def from_list(l: List[dict]) -> List["WidthSpacingTuple"]:
-        out = sorted(list(map(WidthSpacingTuple.from_setting, l)), key=lambda x: x.width_at_least)
+    def from_list(grid_unit: Decimal, l: List[dict]) -> List["WidthSpacingTuple"]:
+        out = sorted(list(map(partial(WidthSpacingTuple.from_setting, grid_unit), l)), key=lambda x: x.width_at_least)
 
         # Check that spacings increase.
-        current_spacing = 0.0
+        current_spacing = Decimal(0)
         for wst in out:
             assert wst.min_spacing >= current_spacing
             current_spacing = wst.min_spacing
         return out
 
 
 class Metal(NamedTuple('Metal', [
+        ('grid_unit', Decimal),
```

Comment:
1) Would the manufacturing grid ever vary per layer?

2) Maybe just to make it clear that the `grid_unit` is a higher-level parameter and not part of the data structure of  `Metal` itself, let's add a comment around it and say that this is not a part of the dict?

Reply:
Yes, it's uncommon, but I have worked in a technology with different manufacturing grids per layer.

Reply:
That said, the way most of this is currently written doesn't support that.

Re: 2, OK, I can add a comment.

Reply:
re. technologies with different manufacturing grids - since we are storing a separate grid_unit in each `Metal` object, might be useful to support an optional `grid_unit` definition in the metal layer itself that gets used if present. This could be filed as an issue as well.